monix-catnap/shared/src/main/scala/monix/catnap/CircuitBreaker.scala

```diff
@@ -0,0 +1,864 @@
+  _maxFailures: Int,
+  _resetTimeout: FiniteDuration,
+  _exponentialBackoffFactor: Double,
+  _maxResetTimeout: Duration,
+  onRejected: F[Unit],
+  onClosed: F[Unit],
+  onHalfOpen: F[Unit],
+  onOpen: F[Unit])
+  (implicit F0: Async[F] OrElse Sync[F], clock: Clock[F]) {
+
+  implicit val F: Sync[F] = F0.unify
+
+  require(_maxFailures >= 0, "maxFailures >= 0")
+  require(_exponentialBackoffFactor >= 1, "exponentialBackoffFactor >= 1")
+  require(_resetTimeout > Duration.Zero, "resetTimeout > 0")
+  require(_maxResetTimeout > Duration.Zero, "maxResetTimeout > 0")
+
+  import monix.catnap.CircuitBreaker._
+  private[this] val stateRef = _stateRef
+
+  /**
+    * The maximum count for allowed failures before opening the circuit breaker.
+    */
+  val maxFailures = _maxFailures
+
+  /** The timespan to wait in the `Open` state before attempting
+    * a close of the circuit breaker (but without the backoff
+    * factor applied).
+    *
+    * If we have a specified [[exponentialBackoffFactor]] then the
+    * actual reset timeout applied will be this value multiplied
+    * repeatedly with that factor, a value that can be found by
+    * querying the [[state]].
+    */
+  val resetTimeout = _resetTimeout
+
+  /** A factor to use for resetting the [[resetTimeout]] when in the
+    * `HalfOpen` state, in case the attempt for `Close` fails.
+    */
+  val exponentialBackoffFactor = _exponentialBackoffFactor
+
+  /** The maximum timespan the circuit breaker is allowed to use
+    * as a [[resetTimeout]] when applying the [[exponentialBackoffFactor]].
+    */
+  val maxResetTimeout = _maxResetTimeout
+
+  /** Returns the current [[CircuitBreaker.State]], meant for
+    * debugging purposes.
+    */
+  val state: F[CircuitBreaker.State] =
+    F.delay(stateRef.get)
+
+  /** Returns a new task that upon execution will execute the given
+    * task, but with the protection of this circuit breaker.
+    */
+  def protect[A](task: F[A]): F[A] =
+    F.suspend(unsafeProtect(task))
+
+  /**
+    * Awaits for this `CircuitBreaker` to be [[CircuitBreaker.Closed closed]].
+    *
+    * This only works if the type class instance used is implementing
+    * [[https://typelevel.org/cats-effect/typeclasses/async.html cats.effect.Async]].
+    *
+    * If this `CircuitBreaker` is already in a closed state, then
+    * it returns immediately, otherwise it will wait (asynchronously) until
+    * the `CircuitBreaker` switches to the [[CircuitBreaker.Closed Closed]]
+    * state again.
+    */
+  def awaitClose(implicit ev: Async[F]): F[Unit] =
+    F.suspend {
+      stateRef.get match {
+        case Closed(_) =>
+          F.unit
+        case Open(_, _, Some(await)) =>
+          await.asInstanceOf[Deferred[F, Unit]].get
+        case HalfOpen(_, Some(await)) =>
+          await.asInstanceOf[Deferred[F, Unit]].get
+        case _ =>
+          // $COVERAGE-OFF$
+          F.raiseError(new APIContractViolationException(
```

Comment:
So, it requires you to know how exactly the CircuitBreaker was constructed in order to understand how this method would behave at runtime? I have a feeling like this is a bad idea.

Reply:
You'll have a hard time building a `CircuitBreaker` instance that reaches this code.

For `IO` and `Task` specifically you'd have to build your own `Sync` instance that's not `Async`. And then you'd have to pass it explicitly in that constructor, otherwise their own `Async[IO]` or `Async[Task]` instances would take precedence.

It's not something that I can see happening. In the wild I'm pretty sure that this branch will never be hit, because it cannot be hit accidentally.

Reply:
Also the gains justifies the means — in this way we keep `CircuitBreaker` working with plain `Sync` data types and then have this little extra just for `Async` data types. It's what parametric polymorphism is all about.

Reply:
An alternative would be to return an `F[Boolean]` instead of throwing an error. I would rather throw the error, because a `Boolean` could be easily ignored.

Reply:
I have added a test to prove it:

https://github.com/monix/monix/pull/744/files#diff-48d1b2884eabc0edcb1c6dbcc95a039cR372

The magic is in `OrElse` definition, which does NOT use subtyping. So even if you have a `Sync[F]` override in scope, if an `Async[F]` is available then it's going to get used.

It's hard to force that error to happen and I don't think it can happen by accident.

Reply:
@alexandru it seems to be trivial with polymorphic code:

```scala
def mkCircuitBreaker[F[_]: Sync] = CircuitBreaker[F].of(...) // boom
```

Reply:
Sure, but such polymorphic code doesn't have access to an `Async[F]` instance for calling `awaitClose` either. And in case the built `CircuitBreaker` object leaks, like in your code, then the error is reasonable imo.

Reply:
I just made a change to also use sub-typing when determining the `Async` instance, so it now looks like this:

```scala
  private def buildAwait: Option[Deferred[F, Unit]] =
    F0.fold(
      // Async
      F => Some(Deferred.unsafeUncancelable[F, Unit](F)),
      // Sync
      {
        case ref: Async[F] @unchecked =>
          Some(Deferred.unsafeUncancelable[F, Unit](ref))
        case _ =>
          None
      })
```

Now this makes it impossible to hit that branch for `IO` and `Task`, unless you force the use of an alternative `Sync[F]` that also risks type class coherence issues.

Also added extra tests for it.

Reply:
> Sure, but such polymorphic code doesn't have access to an Async[F] instance for calling awaitClose either

Polymorphic code can have `Sync[F]` on lower level, but `Async[F]` on higher level. I do it quite often in my own code :)

> I just made a change to also use sub-typing when determining the Async instance

This would be better. Not sure if it would work for transformer instances, however, as those are created on-demand IIRC.

TBH I'd prefer the whole thing to just use `AsyncVar` or something like that, and have `Sync` on creation and `Async` on this particular method.

Reply:
M'kay, let me think about it 🙂 

I would have to hide `State`, to make it a private API, but on the other hand exposing a `Deferred` in `State` is an internals leak too.